### PR TITLE
Handle complex spectral function

### PR DESCRIPTION
**Pull request by gkraberger**
As a first step to the inclusion of the full matrix scheme in this code, I added the treatment of complex spectral functions. Note that when using the G(tau) kernel, this is utterly useless because one can also treat real and imag part independently. Therefore, I implemented the G(iw) kernel (but not the corresponding object to the TauMaxEnt class), where the two effects are mixed. (Of course, an indirect approach where one performs an inverse FT of G(iw) and then FT back the real part of G(tau) to get the G(iw) describing the real part of A would be possible, making it even not strictly necessary in that case either - however, in the full matrix scheme we need also complex treatment.)

I performed a test (it is included and called ``complex_A``) that goes well. 

Does anyone want to have a closer look or should I just merge?

### CODE REVIEW

**Comment by gkraberger** [comment]
For me, the tests work. Can someone with rights to see the jenkins log please give me a hint what goes wrong? Thanks...

**Comment by mzingl** [comment]
23/33 Testing: complex_A
23/33 Test: complex_A
Command: "/Users/jenkins/install/triqs/unstable/gcc/bin/python" "/Users/jenkins/workspace/TRIQS_maxent_PR-2-FR6ZU7VPFI6JYQQ6335X5OTKGE3HZQLYQX2CRGSEHZUN2K5EPO7A/test/python/complex_A.py"
Directory: /Users/jenkins/workspace/TRIQS_maxent_PR-2-FR6ZU7VPFI6JYQQ6335X5OTKGE3HZQLYQX2CRGSEHZUN2K5EPO7A@tmp/build/test/python
"complex_A" start time: Sep 27 08:07 EDT
Output:
----------------------------------------------------------
/Users/jenkins/install/triqs/unstable/gcc/lib/python2.7/site-packages/h5py/__init__.py:72: UserWarning: h5py is running against HDF5 1.10.2 when it was built against 1.8.4, this may cause problems
  '{0}.{1}.{2}'.format(*version.hdf5_built_version_tuple)
Starting on 1 Nodes at : 2018-09-27 08:07:16.356752
Traceback (most recent call last):
  File "/Users/jenkins/workspace/TRIQS_maxent_PR-2-FR6ZU7VPFI6JYQQ6335X5OTKGE3HZQLYQX2CRGSEHZUN2K5EPO7A/test/python/complex_A.py", line 63, in <module>
    assert (np.abs(c2_rp.f() + c2_ip.f() - c2_complex.f())) < 1.e-13
AssertionError
<end of output>
Test time =   0.68 sec
----------------------------------------------------------
Test Failed.
"complex_A" end time: Sep 27 08:07 EDT
"complex_A" time elapsed: 00:00:00

Does this help?

**Comment by mzingl** [comment]
For line 63 - 65 the values are:
3.242917046009097e-11
1.2092641554772854e-06
1.2092641554772854e-06

Also I get this warning:
File "complex_A.py", line 152, in <module>
    result_complex = ml.run()
  File "/home/manuel/Tools/TRIQS_20/lib/python2.7/site-packages/triqs_maxent/maxent_loop.py", line 200, in run
    H[:] = right_side[right_side_slice]
  File "complex_A.py", line 33, in warn_with_traceback
    traceback.print_stack(file=log)
/home/manuel/Tools/TRIQS_20/lib/python2.7/site-packages/triqs_maxent/maxent_loop.py:200: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  H[:] = right_side[right_side_slice]

**Comment by gkraberger** [comment]
Okay, the warning should be dealt with, but that's easy.

I am really very surprised that something that holds up to 10^-10 or better at my machine can fail by > 10^-6 on the test machine. That's not the usual one order of magnitude above the threshold...

**Comment by mzingl** [comment]
That is not only the test machine, but also my ordinary laptop with ubuntu 18

**Comment by gkraberger** [comment]
Good to know. I will test with my ordinary laptop with ubuntu 18 later today and see if I can figure out what makes the difference here...

**Comment by gkraberger** [comment]
I did a simplified test with

```
tau = np.linspace(0, beta, 1)
omega = HyperbolicOmegaMesh(omega_min=-10, omega_max=10, n_points=42)
```

(interestingly, ``n_points=42`` being the lowest number of points where the issue appears - maybe this is an easter egg introduced by some Douglas Adams fans into Ubuntu).

Then, G is just one complex number.

After line 54, I add the following monstrosity
```
print(1,G.view(np.int64))
print(2,G.real.view(np.int64))
print(3,G.imag.view(np.int64))
a = np.dot(K.K, A)
b = np.dot(K.K, A.real)
c = np.dot(K.K, A.imag)
print(4,a.view(np.int64))
print(5,b.view(np.int64))
print(6,c.view(np.int64))
print(7,(a-G).view(np.int64))
print(8,(b-G.real).view(np.int64))
print(9,(c-G.imag).view(np.int64))
print(10,np.abs(a-G).view(np.int64))
print(11,np.abs(b-G.real).view(np.int64))
print(12,np.abs(c-G.imag).view(np.int64))
print(10,(np.abs(a-G)**2).view(np.int64))
print(11,(np.abs(b-G.real)**2).view(np.int64))
print(12,(np.abs(c-G.imag)**2).view(np.int64))
print(13,err.view(np.int64))
print(14,(np.abs(a-G)**2/err**2).view(np.int64))
print(15,(np.abs(b-G.real)**2/err**2).view(np.int64))
print(16,(np.abs(b-G.imag)**2/err**2).view(np.int64))
print(17,(np.abs(b-G.real)**2/err**2 + np.abs(c-G.imag)**2/err**2 - (np.abs(a-G)**2/err**2)).view(np.int64))
print(18,(np.abs(b-G.real)**2/err**2 + np.abs(c-G.imag)**2/err**2 - (np.abs(a-G)**2/err**2)))
```

where the last line gives something on the order of 10^-11 as in the issue.
The ``view(np.int64)`` is there to be able to compare the numbers more easily.

When I run this on my Ubuntu 18 (where I can reproduce the error, just like Manuel), I see
```
1 [-4606363244620462189  4617008914629420410]
2 [-4606363244620462189]
3 [4617008914629420410]
4 [-4606363309517188993  4617008727337586815]
5 [-4606363309517188992]
6 [4617008727337586816]
```
That means that the ``G`` are identical (lines 1-3), but ``np.dot(K.K, A)`` gives a different answer when executed as a complex number (line 4) vs. for real and imaginary part separately (lines 5-6), by one bit.

On the other hand, on my uni computer (Debian),
```
1 [-4606363244620462189  4617008914629420410]
2 [-4606363244620462189]
3 [4617008914629420410]
4 [-4606363309517188993  4617008727337586815]
5 [-4606363309517188993]
6 [4617008727337586815]
```
that problem does not appear.

Now, what I did on the uni computer was, after the definition of a, b and c:
```                                             
bi = b.view(np.int64)                                                   
bi+=1                                                                                                        
ci = c.view(np.int64)                                                   
ci+=1 
```
to artificially introduce the discrepancy there as well.
When I do that, I get *the same* answers as on my Ubuntu laptop from then on, including the 10^-11 discrepancy.

In conclusion
- Ubuntu (or, Lenovo, or whoever is to blame) does a weird bit flip between complex and real numbers for the same matrix multiplication
- This one bit flip has large consequences, causing a difference between numerical zero (i.e. identical bits) and 10^-11 in chi2.

I guess we will have to live with these kind of bit-flip. Therefore we should adjust the threshold (reluctantly), even though (presumably) all kinds of errors could hide with such a rather high threshold.

**Comment by mzingl** [comment]
Thanks Gernot. It works on my computer, but this bothers me. There shouldn't be such a big
difference. Jenkins is still failing in line 63.

**Comment by gkraberger** [comment]
It bothers me extremely, but I fear there is no way of preventing that one-bit difference. I tried different ways of implementing the abs^2 in chi^2 (as real^2+imag^2 or as conj(z) * z) but always got the same effect...

**Comment by mzingl** [comment]
Seems to work now, thanks.
I guess you can merge this pull request.
Best, M
